src/util/rangeutil.hpp

```diff
@@ -117,6 +118,77 @@ sort_by(const Seq& seq, const Proj& proj) {
         });
 }
 
+// Accumulate by projection `proj`
+
+template <
+    typename Seq,
+    typename Proj,
+    typename Value = typename transform_iterator<typename sequence_traits<Seq>::const_iterator, Proj>::value_type
+>
+Value sum_by(const Seq& seq, const Proj& proj, Value base = Value{}) {
+    auto canon = canonical_view(transform_view(seq, proj));
+    return std::accumulate(std::begin(canon), std::end(canon), base);
+}
+
+// Maximum element by projection `proj`
+// - returns an iterator `i` into supplied sequence which has the maximum
+//   value of `proj(*i)`.
+
+template <typename Seq, typename Proj>
+typename sequence_traits<Seq>::iterator
+max_element_by(Seq& seq, const Proj& proj) {
```

Comment:
do we need the const and non-const versions of this? the const version ought to be enough.